---
title: "Tonsil B cell scRNA-seq"
subtitle: "Part 2: Scoring the pySCENIC-derived regullons with AUCell"
author: "Diego A. Espinoza"
date: Last compiled on `r format(Sys.Date(), "%d %B %Y")`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(autodep        = TRUE,
                      echo           = TRUE,
                      error          = FALSE,
                      message        = FALSE,
                      warning        = FALSE)
```

```{r message=FALSE, results='hide', warning=FALSE}
require(magrittr)
require(knitr)
require(dplyr)
require(tidyr)
require(tibble)
require(ggplot2)
require(Seurat)
require(tinytex)
require(cowplot)
require(kableExtra)
require(readxl)
require(reticulate)
require(clusterProfiler)
source("seurat_helper_functions.R")
source("RunAUCell.R")
parent.directory <- "intermediate/"
```


## A. The Seurat object
### A.1 Recap
Recall that we are now performing downstream analysis on 5 independently SCT-normalized datasets that were merged and then integrated using the Seurat v3 method.
```{r intitial-viz, eval = FALSE, fig.width=7, fig.height=7}
load(file = paste0(parent.directory, "sct_named_clusters.RData"))
```

## B. pySCENIC regulons
### B.1 Determining the regulons
The regulons were determined using the pySCENIC pipeline.  We chose to run the pySCENIC pipeline 10 times on the whole dataset. Here we can visualize that many regulons overlap in all 10 runs, so we will only take those that overlap in all 10 runs (we can argue that there is higher confidence in these links).
```{r dtrmn regulons, eval = TRUE}
target_directories <- paste0("from_pyscenic_HPC/w_filter/run_compiled/Bint_pyscenic_regulons_formatted_run_", 1:10, ".csv")
regulon_runs <- lapply(seq_along(target_directories), function(i){
  read.csv(target_directories[i])
}) %>% do.call(rbind, .) %>%
  dplyr::group_by(TF, Targets) %>%
  dplyr::tally(name = "occurrence")
ggplot(regulon_runs, aes(x = occurrence))+
  geom_bar()+
  scale_x_continuous(breaks = 1:10)+
  theme_cowplot()
final_regulon_runs <- regulon_runs %>%
  dplyr::filter(occurrence == 10) %>%
  dplyr::select(TF, Targets)
write.table(x = final_regulon_runs,
            file = paste0(parent.directory, "pyscenic_regulons_merged.csv"),
            sep = ",", quote = FALSE, row.names = FALSE)

```

### B.2 Adding the regulons to the Seurat object
We will then score each of the regulons with the AUCell algorithm and add it to the data as the 'regulons' assay. We add the 'regulons' assay and perform PCA and UMAP on them followed by Louvain clustering as well.
```{r adding regulon data, results="hide", eval = FALSE}
#read in the merged regulons table
regulon_dataframe <- read.csv(file = paste0(parent.directory, "pyscenic_regulons_merged.csv"))

#determine the unique regulons (based on the upstream TF)
my_regulons <- unique(regulon_dataframe$TF)

#prune regulons with less than 10 targets identified (arbitrary cutoff)
small_regulons <- my_regulons[table(regulon_dataframe$TF) < 10]
my_filtered_regulons <- my_regulons[!(my_regulons %in% small_regulons)]
regulon_names <- my_filtered_regulons

#convert the table into a named list of regulons
regulon_list <- lapply(1:length(my_filtered_regulons), function(i){
  regulon_dataframe %>%
    dplyr::filter(TF == regulon_names[i]) %>%
    dplyr::pull(Targets)
})
names(regulon_list) <- paste0(my_filtered_regulons, "-regulon")

#run AUCell scoring algorithm
B.combined <- RunAUCell(B.combined, assay = "RNA", ranking.save = TRUE, genesets = regulon_list, slot = "counts",auc_assay_name = "regulons")

#scale the AUCell regulon values and perform UMAP using custom script. I use 
#a custom `BuildAnnoyUMAP` function which saves the kNN made by the uwot 
#internal algorithm (annoy) instead of creating a a new one with 
#FindNeighbors
DefaultAssay(B.combined) <- "regulons"
B.combined <- ScaleData(B.combined)
B.combined <- BuildAnnoyUMAP(B.combined,
                             assay = "regulons",
                             use_raw = TRUE,
                             slot = "scale.data",
                             n_neighbors = 30,
                             prune_snn = 1/30,
                             metric = "euclidean",
                             reduction_name = "regulon_UMAP",
                             graph_key = "regulon_annoy",
                             reduction_key = "regulonumap_")
B.combined <- FindClusters(B.combined, graph.name = "regulon_annoy_nn_snn", resolution = 0.8)
```

### B.3 Adding the AUC scores for genesets
In order to maintain consistency in scoring genesets, we will also load our other non-regulon geneset AUCs into our Seurat object.
```{r auc genesets, eval = FALSE}
#add geneset AUCs to the Seurat object
my_genesets_dataframe <- ReadGMT("intermediate/genesets.gmt")
my_genesets <- levels(unique(my_genesets_dataframe$term))
my_genesets_names <- my_genesets
geneset_list <- lapply(1:length(my_genesets), function(i){
  my_genesets_dataframe %>%
    dplyr::filter(term == my_genesets_names[i]) %>%
    dplyr::pull(gene)
})
names(geneset_list) <- my_genesets_names

B.combined <- RunAUCell(B.combined, assay = "RNA", ranking.save = FALSE, ranking.key = "ranking", genesets = geneset_list, slot = "counts",auc_assay_name = "genesetAUC")


hm <- clusterProfiler::read.gmt("/Volumes/SanDiskSSD/bar-or_lab/projects/databases/molsigdb/h.all.v7.2.symbols.gmt")
hm_names <- hm %>% dplyr::group_by(term) %>% dplyr::group_keys() %>% dplyr::pull(term)
hm_list <- hm %>%
  dplyr::group_by(term) %>%
  dplyr::group_split() %>%
  lapply(., "[[", 2)
names(hm_list) <- hm_names

B.combined <- RunAUCell(B.combined, assay = "RNA", ranking.save = FALSE, ranking.key = "ranking", genesets = hm_list, slot = "counts",auc_assay_name = "hmAUC")

```

### B.4 Regulon and geneset differential expression
```{r reg and gs DE, eval = FALSE}

diff_exp_regulons <- presto::wilcoxauc(B.combined, seurat_assay = "regulons", group_by = "cluster_names", assay = "counts")
diff_exp_genesets <- presto::wilcoxauc(B.combined, seurat_assay = "genesetAUC", group_by = "cluster_names", assay = "counts")
diff_exp_btms <- presto::wilcoxauc(B.combined, seurat_assay = "btmAUC", group_by = "cluster_names", assay = "counts")
diff_exp_hms <- presto::wilcoxauc(B.combined, seurat_assay = "hmAUC", group_by = "cluster_names", assay = "counts")

save(diff_exp_regulons, diff_exp_genesets, diff_exp_btms, diff_exp_hms, B.combined, file = paste0(parent.directory, "sct_reguloned.RData"))
```



### B.5 Plotting original Louvain clusters on dimensionality reduction results using the regulon and harmony UMAP reductions.
We show the clusters determined by the Louvain algorithm on the harmony_UMAP and regulon_UMAP reductions first.
```{r umap with original clusters, eval = TRUE, fig.width=12}
load(file = paste0(parent.directory, "sct_reguloned.RData"))
DefaultAssay(B.combined) <- "regulons"
dimplot <- DimPlot(B.combined, reduction  = "pca_UMAP",  label = TRUE, label.size = 5, group.by =  "annoy_nn_snn_res.0.6", shuffle = TRUE)+NoLegend()
regulon_dimplot <- DimPlot(B.combined, reduction  = "regulon_UMAP",  label = TRUE, label.size = 5, group.by =  "annoy_nn_snn_res.0.6", shuffle = TRUE)+NoLegend()
dimplot + regulon_dimplot
```

### B.6 Plotting regulon Louvain clusters on dimensionality reduction results using the regulon and harmony UMAP reductions.
We next plot the clusters determined by the Louvain algorithm on the regulon-reduced dataset.
```{r umap with regulon clusters, eval = TRUE, fig.width = 12}
dimplot_regulon_clusters <- DimPlot(B.combined, reduction  = "pca_UMAP", group.by =  "regulon_annoy_nn_snn_res.0.8", label = TRUE, label.size = 5, shuffle = TRUE)+scale_color_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))
regulon_dimplot_regulon_clusters <- DimPlot(B.combined, reduction  = "regulon_UMAP", group.by =  "regulon_annoy_nn_snn_res.0.8", label = TRUE, label.size = 5, shuffle = TRUE)+scale_color_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))
dimplot_regulon_clusters + regulon_dimplot_regulon_clusters
```

### B.7 Alluvial plots 1
```{r alluvial, eval = TRUE, fig.height=6}
B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(RNA, regulon) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  ggplot(aes(x = RNA, y = proport, fill = regulon))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

```

### B.8 Alluvial plots 2
```{r alluvial dos, eval = TRUE, fig.height=6}
B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(regulon, RNA) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  dplyr::filter(grepl("Naive", RNA) | grepl("Marginal", RNA) | grepl("EIF5A", RNA)) %>%
  ggplot(aes(x = regulon, y = proport, fill = RNA))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  coord_cartesian(ylim = c(0,1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(regulon, RNA) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  dplyr::filter(grepl("Memory", RNA)) %>%
  ggplot(aes(x = regulon, y = proport, fill = RNA))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  coord_cartesian(ylim = c(0,1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(regulon, RNA) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  dplyr::filter(grepl("GC", RNA) | grepl("DZ", RNA) | grepl("LZ", RNA) | RNA == "PLCG2") %>%
  ggplot(aes(x = regulon, y = proport, fill = RNA))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  coord_cartesian(ylim = c(0,1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(regulon, RNA) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  dplyr::filter(grepl("ASC", RNA)) %>%
  ggplot(aes(x = regulon, y = proport, fill = RNA))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  coord_cartesian(ylim = c(0,1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

B.combined@meta.data %>%
  select(cluster_names, regulon_annoy_nn_snn_res.0.8) %>%
  set_colnames(c("RNA", "regulon")) %>%
  group_by(regulon, RNA) %>%
  summarize(freq = n()) %>%
  mutate(proport = freq/sum(freq)) %>%
  mutate(regulon = factor(regulon, levels = sort(unique(B.combined$regulon_annoy_nn_snn_res.0.8)))) %>% 
  dplyr::filter(grepl("Activated", RNA)) %>%
  ggplot(aes(x = regulon, y = proport, fill = RNA))+
  geom_col(position = "stack", color = 'black')+
  theme(axis.text.x=element_text(angle = 45, hjust = 1))+
  coord_cartesian(ylim = c(0,1))+
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(11, "Set3")))

```


### B.9 Heatmap of regulon and geneset markers based on RNA clusters
```{r heatmap, eval = TRUE, fig.width=12, fig.height=9}
DoGenesetHeatmap(B.combined, assay = "regulons", diff_exp_results = diff_exp_regulons, group.by = "cluster_names", scale_rows = TRUE, auc_choice = 0.9)
DoGenesetHeatmap(B.combined, assay = "genesetAUC", diff_exp_results = diff_exp_genesets, group.by = "cluster_names", scale_rows = TRUE, auc_choice = 0.8, plot_all = TRUE)
DoGenesetHeatmap(B.combined, assay = "hmAUC", diff_exp_results = diff_exp_hms, group.by = "cluster_names", scale_rows = TRUE, auc_choice = 0.8, plot_all = FALSE)

```


### B.10 Heatmap of markers based on regulon clusters
```{r heatmap 2, eval = TRUE, fig.width=12, fig.height=9}
Idents(B.combined) <- "regulon_annoy_nn_snn_res.0.8"
diff_exp_regulons_by_regulons <- presto::wilcoxauc(B.combined, seurat_assay = "regulons", group_by = "regulon_annoy_nn_snn_res.0.8", assay = "counts")
DoGenesetHeatmap(B.combined, assay = "regulons", diff_exp_results = diff_exp_regulons_by_regulons, group.by = "regulon_annoy_nn_snn_res.0.8", scale_rows = TRUE, auc_choice = 0.8, plot_all = FALSE)
```


